src/comparator/field_comparator.py

```diff
@@ -186,8 +186,17 @@ def compare(self):
                     message=msg,
                     change_type=ChangeType.MAJOR,
                 )
+        # 6. Check the proto3_optional state of the field.
+        if self.field_original.proto3_optional != self.field_update.proto3_optional:
+            self.finding_container.addFinding(
+                category=FindingCategory.FIELD_PROTO3_OPTIONAL_CHANGE,
+                proto_file_name=self.field_update.proto_file_name,
+                source_code_line=self.field_update.source_code_line,
+                message=f"Proto3 optional state of an existing field `{self.field_original.name}` is changed.",
+                change_type=ChangeType.MAJOR,
```

Comment:
Isn't it only a minor change if changing from required to optional?

Reply:
Sure, I was under impression in the [proto](https://source.corp.google.com/piper///depot/google3/net/proto2/proto/descriptor.proto;l=277) that the proto3_optional always belong to a `oneof`, and the fields in oneof should always be optional ([aip](https://google.aip.dev/apps/2717)) so from optional to required is considered **breaking**.  While based on the comments from @chrisdunelm below, the proto3_optional can be attached to any field, then yes, `non-optional` to `optional` is not breaking. Thank you

Reply:
To confirm, the existing field is moved from one one-of to another one-of is not breaking, for example oneof renaming. At least from Typescript, Python micro-gen, this change won't affect the client lib generation, but double check here. Thanks